### PR TITLE
feat: SelfClientsUseCase [AR-980]

### PR DESCRIPTION
**Pull request by MohamadJaara**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [ ] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

[AR-980]

### Solutions

implemnting SelfClientsUseCase

Needs releases with:

- [ ] GitHub link to other pull request

### Testing
WIP

----
#### PR Post Merge Checklist for internal contributors

 - [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

[AR-980]: https://wearezeta.atlassian.net/browse/AR-980?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientModel.kt [hunk 1]
```diff
@@ -11,6 +11,11 @@ data class RegisterClientParam(
     val capabilities: List<ClientCapability>?,
 )
 
+data class DeleteClientParam(
+    val password: String,
+    val clientId: ClientId /* = com.wire.kalium.logic.data.id.PlainId */
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientModel.kt:16]: Is this common practice? To write a comment everywhere you use a typealias?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/remote/ClientRemoteRepository.kt [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    override suspend fun fetchClientInfo(clientId: ClientId): Either<CoreFailure, Client> {
+        val response = clientApi.fetchClientInfo(clientId.value)
+        return if (response.isSuccessful()) {
+            Either.Right(clientMapper.fromClientResponse(response.value))
+        } else {
+            handleFailedApiResponse(response)
+        }
+    }
+
+    override suspend fun fetchSelfUserClient(): Either<CoreFailure, List<Client>> {
```
Comment (typfel) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/remote/ClientRemoteRepository.kt:58]:
```suggestion
    override suspend fun fetchSelfUserClients(): Either<CoreFailure, List<Client>> {
```

**Review by typfel — CHANGES_REQUESTED** [review]

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/ClientScope.kt [hunk 3]
```diff
@@ -7,7 +7,7 @@ import com.wire.kalium.logic.data.prekey.PreKeyMapper
 class ClientScope(
     clientRepository: ClientRepository,
     proteusClient: ProteusClient,
-    preKeyMapper: PreKeyMapper
 ) {
-    val register: RegisterClientUseCase = RegisterClientUseCaseImpl(clientRepository)
+    // TODO : get() ?
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/client/ClientScope.kt:11]: I guess?

File: logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientRepository.kt [hunk 4]
```diff
@@ -1,11 +1,49 @@
...
+class ClientDataSource(
+    private val clientRemoteRepository: ClientRemoteRepository,
+    private val clientRegistrationStorage: ClientRegistrationStorage
+) : ClientRepository {
+    override suspend fun registerClient(param: RegisterClientParam): Either<CoreFailure, Client> {
+        return clientRemoteRepository.registerClient(param)
+    }
+
+    override suspend fun persistClientId(clientId: ClientId): Either<CoreFailure, Unit> {
+        clientRegistrationStorage.registeredClientId = clientId.value
```
Comment (vitorhugods) [inline logic/src/commonMain/kotlin/com/wire/kalium/logic/data/client/ClientRepository.kt:28]: This makes me think... we are never catching DB Exceptions.

Maybe in the future we could wrap repositories' operations that touch the DB to catch them and return do something like:

```kotlin
fun wrapDBOperation(action: () -> T): Either<CoreFailure, T> {
    return try {
        Either.Right(action())
    } catch(dbException: SQLException){
        Either.Left(CoreFailure.MrUserIDontFeelSoGood)
    }
}
```

**Review by vitorhugods — APPROVED** [review]

**Review by typfel — APPROVED** [review]

**Pull request merged**